test: Add mempool_updatefromblock.py

hebasto opened the pull request:
This PR adds a new test for mempool update of transaction descendants/ancestors information (count, size) when transactions have been re-added from a disconnected block to the mempool.

It could be helpful for working on PRs like #17925, #18191.

hebasto committed:
test: Add mempool_updatefromblock.py

JeremyRubin commented:
Concept ACK.

I don't see any reason not to accept this test as is; but it could be nice to:

1) Make it a bit more generic (e.g., specify a param N rather than writing out each transaction)
2) Test some more complicated transaction graphs

hebasto commented:
Updated 7ed5558042889e08cee941e697b7547213877e9b -> cfaa79b9f4d0c220c220ae0d8a6825581aa89133 ([pr18485.01](https://github.com/hebasto/bitcoin/commits/pr18485.01) -> [pr18485.02](https://github.com/hebasto/bitcoin/commits/18485.02), [diff](https://github.com/hebasto/bitcoin/compare/pr18485.01..pr18485.02)).

@JeremyRubin 
> 1. Make it a bit more generic (e.g., specify a param N rather than writing out each transaction)

Done.

> 2. Test some more complicated transaction graphs

Done.

JeremyRubin commented:
Looks great :)

hebasto commented:
Updated cfaa79b9f4d0c220c220ae0d8a6825581aa89133 -> b9156fd3d10dcf5cf18a4728ae42f39826a00586 ([pr18485.02](https://github.com/hebasto/bitcoin/commits/pr18485.02) -> [pr18485.03](https://github.com/hebasto/bitcoin/commits/18485.03), [diff](https://github.com/hebasto/bitcoin/compare/pr18485.02..pr18485.03)):

- fixed some `flake8` style warnings.

ariard commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -0,0 +1,121 @@
+"""Test mempool descendants/ancestors information update.
+
+Test mempool update of transaction descendants/ancestors information (count, size)
+when transactions have been re-added from a disconnected block to the mempool.
```
Maybe hints directly to `UpdateTransactionsFromBlock`

hebasto replied:
> Maybe hints directly to `UpdateTransactionsFromBlock`

This could be more appropriate for unit tests rather for functional ones, no?

ariard replied:
I'm don't have a strong opinon, but sometimes for newcomers is great to lead them directly to what part of the codebase is tested.

ariard commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -0,0 +1,121 @@
+            - has N-K descendants (including this one), and
+            - has K+1 ancestors (including this one)
+
+        More details: https://en.wikipedia.org/wiki/Tournament_(graph_theory)
```
A tournament can be a direct cyclic graph? But utxo graph has to be a direct acyclic graph and should be forbid in the mempool, for preciseness, you may indicate it generates a subset of tournament.

hebasto replied:
I think "acyclic" property is implied in this case :)

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18485#issuecomment-619077646).

ariard commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -0,0 +1,121 @@
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', end_address='', fee=Decimal(0.00100000)):
```
How people should use `start_input_txid` and `end_address` ? Do you envision having different helpers each with a specific topology and being able to connect them to test different mempool entry points ? Like a N-fan-out connected to a linear chain ?

hebasto replied:
> Do you envision having different helpers each with a specific topology and being able to connect them to test different mempool entry points ? Like a N-fan-out connected to a linear chain ?

You read my mind :)

ariard replied:
Great, will review next ones too!

ariard commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -0,0 +1,121 @@
+                inputs = []
+                inputs_value = 0
+                for j, tx in enumerate(tx_id[0:i]):
+                    vout = i - j - 1
```
"Transaction N is children of every previous transactions at their output N-1"

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18485#issuecomment-619077646).

ariard commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -0,0 +1,121 @@
+            # Prepare outputs.
+            tx_count = i + 1
+            if tx_count < size:
+                n_outputs = size - tx_count
```
"Transaction N is ancestor of every transactions[N+1...size]"

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18485#issuecomment-619077646).

ariard reviewed (approved):
ACK b9156fd

Cool work! It's mostly suggestions to make test clearer, feel free to ignore them.

hebasto commented:
Updated b9156fd3d10dcf5cf18a4728ae42f39826a00586 -> 8098dea06944f9de8b285f44958eb98761f133ee ([pr18485.03](https://github.com/hebasto/bitcoin/commits/pr18485.03) -> [pr18485.04](https://github.com/hebasto/bitcoin/commits/pr18485.04), [diff](https://github.com/hebasto/bitcoin/compare/pr18485.03..pr18485.04)):

- addressed @ariard's [comments](https://github.com/bitcoin/bitcoin/pull/18485#pullrequestreview-399569822):

> ACK [b9156fd](https://github.com/bitcoin/bitcoin/commit/b9156fd3d10dcf5cf18a4728ae42f39826a00586)
> 
> Cool work! It's mostly suggestions to make test clearer, feel free to ignore them.

ariard commented:
ACK 8098dea
